### PR TITLE
add test for pagination pages

### PR DESCRIPTION
**Pull request by werwty**

### CODE REVIEW

File: pypi/projects.py [hunk 1]
```diff
@@ -23,3 +23,47 @@
...
+
+
+async def get_pages(
+    num_pages: int, queue, url: str = defaults.API_URL, timeout: int = defaults.TIMEOUT
+) -> List:
+    queue = asyncio.Queue()
+
+    # Create X Consumers
+    consumers = [
+        asyncio.ensure_future(_consume_page(queue)) for i in range(defaults.CONSUMERS)
```
Comment (cooperlees) [inline pypi/projects.py]: What was the error if you added raw _consume_pages here? I thought they should be queue-able
- Futures will also work.

File: pypi/projects.py [hunk 2]
```diff
@@ -23,3 +23,47 @@
...
+    # cancel all consumers after queue.join() is complete
+    for consumer in consumers:
+        consumer.cancel()
+
+
+async def _consume_page(queue):
+    ret = []
+    while True:
+        item = await queue.get()
+        ret.append(await item)
```
Comment (cooperlees) [inline pypi/projects.py]: Can we swap the name of item to page_get ? Or something a little more descriptive than item

File: pypi/tests/base.py [hunk 3]
```diff
@@ -18,21 +20,28 @@ def test_latest(self) -> None:
...
-        js = self.loop.run_until_complete(journals.since(latest['last_serial']))
+        js = self.loop.run_until_complete(journals.since(latest["last_serial"]))
         self.assertTrue(len(js), 5000)
 
 
 class TestProjects(unittest.TestCase):
 
     def setUp(self):
         self.loop = asyncio.get_event_loop()
+        self.queue = asyncio.Queue(loop=self.loop, maxsize=10)
```
Comment (cooperlees) [inline pypi/tests/base.py:30]: Just a FYI = self.loop would of been selected by default as it's the default loop - I like explicit, but just letting you know.

File: pypi/tests/base.py [hunk 4]
```diff
@@ -18,21 +20,28 @@ def test_latest(self) -> None:
...
+        pulp = self.loop.run_until_complete(projects.get_detail("pulpcore"))
         self.assertTrue("3.0.0a1.dev0" in pulp["releases"])
         self.assertEqual(275959, pulp["last_serial"])
 
     def test_since(self) -> None:
         since_projects = self.loop.run_until_complete(projects.since(275959)),
         self.assertTrue(since_projects)
+
+    def test_paginate(self) -> None:
+        i: int = 9999
```
Comment (cooperlees) [inline pypi/tests/base.py]: Can we add a comment to state we're using a large number in order to find the actual last page.

Also, for an int, as it's a built in type, Python can 'type infer' it, so it's not needed, but once again I have nothing against explicit.

File: pypi/tests/base.py [hunk 5]
```diff
@@ -18,21 +20,28 @@ def test_latest(self) -> None:
...
+        page = self.loop.run_until_complete(projects.get_page(i))
+        last_page = parse_qs(urlparse(page["links"]["previous_page"]).query)["page"][0]
```
Comment (cooperlees) [inline pypi/tests/base.py]: We should probably provide a helper function somewhere for this as it will be a common operation

**Review by cooperlees — CHANGES_REQUESTED** [review]
Just some small nits - Message me on IRC to talk about anything 💃

File: pypi/projects.py (hunk 1 above)
Comment (werwty) [reply pypi/projects.py]: Without ensure_futures execution of `await queue.join()` would hang indefinitely.

**Comment by werwty** [comment]
Thanks for all your help @cooperlees! :cheese: :clinking_glasses: :tada:

**Review by cooperlees — APPROVED** [review]

**Pull request merged**